# Customer Segmentation using the RFM Methodology

## Introduction


The RFM model is a dynamic framework that unveils the intricate tapestry of customer behaviors and preferences. In today's rapidly evolving market, understanding customers' distinct traits is paramount. RFM stands as a beacon, illuminating the path to precision in marketing and service strategies.

This project uses the RFM model to segment customers to enable the business find out its best customers.

In [2]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')


In [3]:
#LOading the data
df = pd.read_csv(r'C:\Users\Tette\OneDrive\Documents\DATA SETS\online_retail.csv')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [6]:
df.shape
df.columns.values

array(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'], dtype=object)

In [5]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [13]:
data = df.dropna()

In [14]:
customer_country=data[['Country','CustomerID']].drop_duplicates()
customer_country.groupby(['Country'])['CustomerID'].aggregate('count').reset_index().sort_values('CustomerID', ascending=False)

Country  CustomerID
35        United Kingdom        3950
14               Germany          95
13                France          87
30                 Spain          31
3                Belgium          25
32           Switzerland          21
26              Portugal          19
18                 Italy          15
12               Finland          12
1                Austria          11
24                Norway          10
23           Netherlands           9
0              Australia           9
9                Denmark           9
6        Channel Islands           9
7                 Cyprus           8
31                Sweden           8
19                 Japan           8
25                Poland           6
33                   USA           4
5                 Canada           4
36           Unspecified           4
17                Israel           4
15                Greece           4
10                  EIRE           3
22                 Malta           2
34  United Arab Emirates           2
2                Bahrain           2
8         Czech Republic           1
21             Lithuania           1
20               Lebanon           1
27                   RSA           1
28          Saudi Arabia           1
29             Singapore           1
16               Iceland           1
4                 Brazil           1
11    European Community           1

In [18]:
data = data[(data['Quantity'] > 0)]
data.head()
data.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [20]:
#Total Prices of items sold

data['TotalPrice'] = data['Quantity'] * data['UnitPrice']
print(data)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                InvoiceDate  UnitPrice  CustomerID         Country  TotalPr

In [23]:
data = data[data['TotalPrice'] > 0]
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0       2010-12-01 08:26:00       2.55     17850.0  United Kingdom       15.30  
1       2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
2       2010-12-01 08:26:00       2.75     17850.0  United Kingdom       22.00  
3       2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
4       2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
...                     ...        ...         ...             ...         ...  
541904  2011-12-09 12:50:00       0.85     12680.0          France       10.20  
541905  2011-12-09 12:50:00       2.10     12680.0          France       12.60  
541906  2011-12-09 12:50:00       4.15     12680.0          France       16.60  
541907  2011-12-09 12:50:00       4.15     12680.0          France       16.60  
541908  2011-12-09 12:50:00       4.95     12680.0          France       14.85  

[397884 rows x 9 columns]

In [25]:
data['InvoiceDate'].max() 

'2011-12-09 12:50:00'

In [27]:
#Setting the data to ajust for recency

import datetime as dt
NOW = dt.datetime(2011,12,10)
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

Data has been prepared for the segmentation. Missing values were deleted as customers with no ID will be irrelevant for the purpose of the analysis. The last transaction date recorded was 9th December 2011, we can adjust for the recency to 2011-12-10 to reference all invoice dates by NOW. 

The number of days between the most recent transaction date and the current date will be calculated as recency, the number of invoices or transactions made by each customer as frequency and the total monetary value spent by each customer will be included as well.

### Segmentation of Customers
The data is grouped by 'CustomerID' aimed at aggregating information for each unique customer. 

In [33]:
rfmTable = data.groupby('CustomerID').agg({'InvoiceDate': lambda x: (NOW - x.max()).days, 'InvoiceNo': lambda x: len(x), 'TotalPrice': lambda x: x.sum()})
rfmTable['InvoiceDate'] = rfmTable['InvoiceDate'].astype(int)
rfmTable.rename(columns={'InvoiceDate': 'recency', 
                         'InvoiceNo': 'frequency', 
                         'TotalPrice': 'monetary_value'}, inplace=True)

print(rfmTable)

            recency  frequency  monetary_value
CustomerID                                    
12346.0         325          1        77183.60
12347.0           2        182         4310.00
12348.0          75         31         1797.24
12349.0          18         73         1757.55
12350.0         310         17          334.40
...             ...        ...             ...
18280.0         277         10          180.60
18281.0         180          7           80.82
18282.0           7         12          178.05
18283.0           3        756         2094.88
18287.0          42         70         1837.28

[4338 rows x 3 columns]


Next, Quartiles will be used to divide a dataset into four equal parts, providing a clear understanding of the distribution of the data. In this analysis, this means dividing customers into four segments based on their recency, frequency, and monetary value scores.

In [37]:
quantiles = rfmTable.quantile(q=[0.25,0.5,0.75])
quantiles = quantiles.to_dict()
print(quantiles)

{'recency': {0.25: 17.0, 0.5: 50.0, 0.75: 141.75}, 'frequency': {0.25: 17.0, 0.5: 41.0, 0.75: 100.0}, 'monetary_value': {0.25: 307.41499999999996, 0.5: 674.4849999999999, 0.75: 1661.7400000000002}}


These quartiles represent key thresholds across recency, frequency, and monetary value metrics derived from an RFM analysis. They segment customers into four equal parts, illustrating varying degrees of engagement and value. For recency, 25% of customers made a purchase within 17 days or less (Q1), while 50% did so within 50 days (Q2), and 75% within 142 days (Q3). Regarding frequency, 25% of customers made 17 purchases or fewer (Q1), 50% made 41 purchases (Q2), and 75% made 100 purchases or fewer (Q3). In terms of monetary value, 25% of customers spent $307.41 or less (Q1), 50% spent $674.48 or less (Q2), and 75% spent $1661.74 or less (Q3). These quartiles serve as benchmarks to categorize customers into distinct segments, aiding businesses in targeted strategies and resource allocation based on differing customer engagement and value levels.

To find our best customers, we need customers with the lowest recency, highest frequency and best monetary amounts.

In [40]:
segmented_rfm = rfmTable

In [38]:
#Setting functions to assign scores based on the quartiles a customer may fall into.
def RScore(x,p,d):
    if x <= d[p][0.25]:
        return 1
    elif x <= d[p][0.50]:
        return 2
    elif x <= d[p][0.75]: 
        return 3
    else:
        return 4
    
def FMScore(x,p,d):
    if x <= d[p][0.25]:
        return 4
    elif x <= d[p][0.50]:
        return 3
    elif x <= d[p][0.75]: 
        return 2
    else:
        return 1

In [43]:
#Creating new columns for the scores generated using the function above

segmented_rfm['r_quartile'] = segmented_rfm['recency'].apply(RScore, args=('recency',quantiles,))
segmented_rfm['f_quartile'] = segmented_rfm['frequency'].apply(FMScore, args=('frequency',quantiles,))
segmented_rfm['m_quartile'] = segmented_rfm['monetary_value'].apply(FMScore, args=('monetary_value',quantiles,))
segmented_rfm

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   
...             ...        ...             ...         ...         ...   
18280.0         277         10          180.60           4           4   
18281.0         180          7           80.82           4           4   
18282.0           7         12          178.05           1           4   
18283.0           3        756         2094.88           1           1   
18287.0          42         70         1837.28           2           2   

            m_quartile  
CustomerID              
12346.0              1  
12347.0              1  
12348.0              1  
12349.0              1  
12350.0              3  
...                ...  
18280.0              4  
18281.0              4  
18282.0              4  
18283.0              1  
18287.0              1  

[4338 rows x 6 columns]

A column is now created called 'RFMScore' in the DataFrame segmented_rfm by combining the values from the columns 'r_quartile', 'f_quartile', and 'm_quartile' after converting them to strings. This is done so we can easily sort out '111' which would be our ideal customers.

In [45]:
segmented_rfm['RFMScore'] = segmented_rfm.r_quartile.map(str) + segmented_rfm.f_quartile.map(str) + segmented_rfm.m_quartile.map(str)
segmented_rfm.head()

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12347.0           2        182         4310.00           1           1   
12348.0          75         31         1797.24           3           3   
12349.0          18         73         1757.55           2           2   
12350.0         310         17          334.40           4           4   

            m_quartile RFMScore  
CustomerID                       
12346.0              1      441  
12347.0              1      111  
12348.0              1      331  
12349.0              1      221  
12350.0              3      443

Now we can find a list of our best customers as:


In [46]:
segmented_rfm[segmented_rfm['RFMScore']=='111'].sort_values('monetary_value', ascending=False)

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
14646.0           1       2076       280206.02           1           1   
18102.0           0        431       259657.30           1           1   
17450.0           8        337       194550.79           1           1   
14911.0           1       5675       143825.06           1           1   
14156.0           9       1400       117379.63           1           1   
...             ...        ...             ...         ...         ...   
17813.0          14        371         1770.88           1           1   
13184.0          14        181         1701.29           1           1   
16775.0          10        158         1695.66           1           1   
16813.0           8        449         1692.98           1           1   
16115.0           9        284         1667.97           1           1   

            m_quartile RFMScore  
CustomerID                       
14646.0              1      111  
18102.0              1      111  
17450.0              1      111  
14911.0              1      111  
14156.0              1      111  
...                ...      ...  
17813.0              1      111  
13184.0              1      111  
16775.0              1      111  
16813.0              1      111  
16115.0              1      111  

[440 rows x 7 columns]

## Conclusion
In analyzing a dataset of 397,884 customers, we've identified a select group of 444 customers who stand out as the best performers across recency, frequency, and monetary value metrics based on RFM analysis. These 444 customers represent the top-tier segment, demonstrating superior engagement, frequent transactions, and substantial monetary contributions to the business compared to the broader customer base. Their exceptional behavior in terms of recent purchases, high frequency, and substantial spending distinguishes them as the most valuable and engaged customers within this dataset. Understanding and catering to the preferences and needs of these top 444 customers can significantly impact marketing strategies, loyalty programs, and resource allocation, fostering stronger customer relationships and potentially driving substantial revenue growth.